src/ServiceResolver/Factory/ServiceResolverFactory.php

```diff
@@ -1,16 +1,19 @@
 <?php
 
-namespace Bauhaus;
+namespace Bauhaus\ServiceResolver\Factory;
 
+use Bauhaus\ServiceResolver;
 use Bauhaus\ServiceResolver\Resolver;
-use Bauhaus\ServiceResolver\Resolvers\CircularDependencyDetector;
-use Bauhaus\ServiceResolver\Resolvers\Discoverer;
-use Bauhaus\ServiceResolver\Resolvers\MemoryCache;
-use Bauhaus\ServiceResolver\Resolvers\ServiceDefinitionContainer;
-use Bauhaus\ServiceResolver\ServiceDefinition;
-use Bauhaus\ServiceResolver\ServiceDefinitionLoader;
-use Bauhaus\ServiceResolver\ServiceDefinitionLoaderException;
+use Bauhaus\ServiceResolver\Resolvers\CircularDependencyDetector\CircularDependencyDetector;
+use Bauhaus\ServiceResolver\Resolvers\Discoverer\Discoverer;
+use Bauhaus\ServiceResolver\Resolvers\MemoryCache\MemoryCache;
+use Bauhaus\ServiceResolver\Resolvers\DefinitionCollection;
+use Bauhaus\ServiceResolver\Definition;
+use Bauhaus\ServiceResolverOptions;
 
+/**
+ * @internal
```

Comment:
Is this class really `internal`? I thought this was the expected way for users to build a ServiceResolver.

Reply:
I agree and, previously, this class was actually in the root namespace. So, it would be possible to access it from the out world, but I didn't like for some reasons.

I have another solution suggestion: 90a676d. This introduces the method `build` to `ServiceResolverOptions` which returns a `ServiceResolver` instance. In this way you do not need to know the details of how it is created. What do you think?

Reply:
Could you list some reasons for not exposing the Factory? :)

As I [commented below](https://github.com/bauhausphp/service-resolver/pull/10#discussion_r602883655), I found this `Options->build()` surprising. An alternative could be providing this method as `ServiceResolver::buildFrom(ServiceResolverOptions)`.

Reply:
This factory is just a refactoring of how one class is created, but, in my vision, it is purely implementation detail that is not relevant for the interface.

As I also [commend below](https://github.com/bauhausphp/service-resolver/pull/10#discussion_r602884509) 😄, the option class is following the _Builder Design Pattern_ and a renaming might be needed.

Anyway, your suggestion is indeed a good alternative. I will read a bit more about the _Builder Design Pattern_ before deciding the approach. So, I will let this discussion open.

Reply:
So, I read about the _Builder Design Pattern_. My conclusion is: `ServiceResolverOptions` was somewhat close to its definition, but not exactly the same. To avoid any misinformation or confusion, I decided to abort this approach.

In the end, I followed your suggestion to have the method `ServiceResolver::build` and I also renamed the factory to `ResolverChainFactory` which returns the `Resolver` interface. I liked because now all parts look "in peace" with each other.

You can check it out here: 08442a0.